control_plane/src/storage.rs

```diff
@@ -235,15 +234,39 @@ impl PageServerNode {
             args.extend(["-c", config_override]);
         }
 
-        let mut cmd = Command::new(self.env.pageserver_bin()?);
-        let mut filled_cmd = fill_rust_env_vars(cmd.args(&args).arg("--daemonize"));
-        filled_cmd = fill_aws_secrets_vars(filled_cmd);
-
-        if !filled_cmd.status()?.success() {
-            bail!(
-                "Pageserver failed to start. See console output and '{}' for details.",
-                datadir.join("pageserver.log").display()
-            );
+        let pageserver_log_path = datadir.join("pageserver.log");
+        let pagesesrver_log_file = fs::OpenOptions::new()
+            .create(true)
+            .write(true)
+            .truncate(true)
+            .open(&pageserver_log_path)
+            .with_context(|| {
+                format!("Failed to open pageserver log file {pageserver_log_path:?} for logging")
+            })?;
+
+        let background_command = subprocess::Exec::cmd(self.env.pageserver_bin()?)
+            // TODO kb wrong, need to point at ps file? also set curdir?
+            .stdout(subprocess::Redirection::File(pagesesrver_log_file))
+            .stderr(subprocess::Redirection::Merge)
+            .args(&args)
+            // do not wait for the subprocess to exit when it gets dropped: we write its pid into a file
+            // to retrieve and manage on another CLI restart
+            .detached();
+        let filled_cmd = fill_aws_secrets_vars!(fill_rust_env_vars!(background_command));
+
+        // by default, `subprocess::Popen` waits for process to complete on drop
+        let mut pageserver_subprocess = filled_cmd.popen().with_context(|| {
+            format!("Pageserver failed to start. See console output and {pageserver_log_path:?} for details.")
+        })?;
+        let pageserver_pid = pageserver_subprocess
+            .pid()
+            .context("Pageserver subprocess got no pid")?;
+        let pid_file_path = self.pid_file();
+        if pid_file_path.exists() {
+            let pid_string = std::fs::read_to_string(&pid_file_path)
+                .with_context(|| format!("Failed to read pageserver pid file {pid_file_path:?}"))?;
+            anyhow::ensure!(pageserver_pid.to_string() == pid_string,
+                "Pageserver subprocess pid {pageserver_pid} does not match pageserver pid file pid {pid_string}");
         }
```

Comment:
Race condition: The pageserver might be slower than this process, and it might not have created the pid file yet.

Reply:
Reworked the way pidfile/lockfile is treated now.